### DISCUSSION TITLE
Key and Data sizes with DUPSORT

### DISCUSSION THREAD

**Discussion by josephnicholas** [erthink/libmdbx#358] [Q&A] [answered]
Hello @leo-yuriev ,

We have an issue before previously in LMDB when handling post quantum signature schemes. Since the keys was so big, there was a limitation in LMDB. When a db is created with `DUPSORT` flag enabled the key and data values would have the same size of 0 - 511 bytes.

- So my question would be does `libmdbx` inherit the LMDB key size limitation or is it different?

**Comment by erthink** [comment, marked as answer]
Since 2020 _libmdbx_ supports keys and dupsort values **2 times longer than _LMDB_** for the same DB page size. Please use [`mdbx_limits_keysize_max()`](https://libmdbx.dqdkfa.ru/group__c__statinfo.html#gab833240078c4d8effadf1b7bf3a7a44f), [`mdbx_limits_valsize4page_max()`](https://libmdbx.dqdkfa.ru/group__c__statinfo.html#ga52ea51805876cb1e3289b18451897337) and [`mdbx_limits_pairsize4page_max()`](https://libmdbx.dqdkfa.ru/group__c__statinfo.html#gaca5c03fb63421882797a56854aa24098) to get the exactly values.

**Comment by josephnicholas** [comment]
Thank you @leo-yuriev .

Just a follow-up use case, so I currently have some data:

`<amount>[32 bytes] : <cryptographic key>[4.96KB]`

So what I want to do is create a database which enables DUP and DUPSORT, basically same as what the LMDB flags here does:
https://github.com/LMDB/bitmonero/blob/d0d76f771ac318046306cd2ef40e67d598d7429d/src/blockchain_db/lmdb/db_lmdb.cpp#L1400

So currently, as I've said, LMDB doesn't work for me, since when `DUPSORT` is enabled, I am limited by this:
https://github.com/LMDB/bitmonero/blob/d0d76f771ac318046306cd2ef40e67d598d7429d/external/db_drivers/liblmdb/mdb.c#L640-L658

So, in what you said about building a custom page size, I am not really familiar with `libmdbx` quite yet, am I correct in assuming that, for example:

1. I created a DB with Page Size 16K, with same FLAGS as my link above.
2. Knowing that the I can have 1/4 of the page size as my value size, which is ~8KB.
3. I can fit the `[32 byte key] : [4.96 KB value]`

Am I correct on this assumption, also this is for the current implementation.

**Comment by erthink** [comment]
For 4096-bytes dupsort values, you need a database page size of at least 32K, since the max key size is _slightly smaller_ than 1/4 the page size. I. e. with page size 8192 you will get about 2000 maximum key size, and about 4000 with 16384. So it seems only 32768 page size will be enough for your case.

---

Just in case, I note that the support of long keys is always quite expensive. All databases have similar restrictions (e.g. ORA-01450). And in libmdbx and LMDB case the DUPSORT feature spreads the same restrictions to length of values.

Therefore, it is obvious suggestion: just avoid large keys both for reasons space and performance. In your case one of solution is the "external key", i.e. use another key-value space where key is a sequence (e.g. [`imdbx_dbi_sequence()`](https://github.com/leo-yuriev/libmdbx/blob/1ab33333bcde54cce99d78f67fa59ff7eea42331/mdbx.h#L2794-L2796)) and value is "signature data", and then store only sequence-key in the main key-value space.

So, it may be [libfpta](https://github.com/leo-yuriev/libfpta) (soon to be re-licensed under apache 2.0) will be more useful in your case, since it support secondary indexes out-of-the box, including unordered (i.e. hashe-based indexed).

With `libfpta` you should create more complex database schema than key-value. For instance I am suggest row columns:
 1. primary key = unique uint64.
 2. 32-bytes binary "amount" column, with secondary non-unique index.
 3. 4096-bytes binary "signature" column, with with secondary unordered index.

This solution needs to be discussed and refined, but it is likely that it will be better in all respects (both performance and database size).

**Comment by josephnicholas** [comment]
@leo-yuriev Thank you for discussing into detail your suggestions. I really do appreciate the help. I have a question though I hope you don't mind, since maybe I can't use `libfpta` yet for some licensing reason right?

So can you please discuss more on how would the "external key" on another "key-value" space help? Does this mean I have to create a DB then store on that other DB just the [4096]"signature data" and the main just the "sequence-key" on the main? But would it also be hard to access them?

**Comment by erthink** [comment]
A "key-value space" means a "database" in the LMDB' terminology (I think "space" is more suitable, since "database" make confusion with "a whole database as the set of a spaces" that LMDB calls "environment").

The "external key" concept is very similar to "foreign key" from SQL, but in key-value DB you need done the same job by yourself. In general, you should split your key-value items into two LMDB's databases and link them by rowid (i.e. by the row numbers). For instance:
 - create database/space "amount" where amount is the key and rowid is the data;
 - create database/space "cryptographic key" where rowid is the key;
 - for access to "cryptographic key" you need one more lookup, etc;

Despite the overhead, this scheme works (mostly) faster in cases like yours, because b-tree is spared long keys. The actual result depends on the size of the keys and the number of duplicates. However, this scheme has more stable and predictable behavior even in extreme cases.

**Comment by josephnicholas** [comment]
Thanks @leo-yuriev very well explained, I will try to implement this based on what I understood.

Anyway, the use case only happens for a specific data, some are okay with LMDB, only this use case is really not handled. So is it advisable to implement LMDB on some and implement libmdbx on this specific one, or should I rewrite everything with `libmdbx`?. I know this actually to be answered by me(since this is my past time project), but I need your advice since you are more of an expert than I am on this area.

Thanks.

**Comment by erthink** [comment]
libmdbx can give you more, but only yourself can decide how useful these additional features are to you. I suggest you examine with the [list of differences](https://github.com/leo-yuriev/libmdbx#improvements-over-lmdb) libmdbx from LMDB and read the information in the [mdbx.h](https://github.com/leo-yuriev/libmdbx/blob/master/mdbx.h).
